### PR TITLE
WFCORE-648 Ear exclusions cascaded to sub-deployments

### PR DESCRIPTION
**Pull request by bgaisford**
Within jboss-deployment-structure.xml, add a convenient way to cascade ear deployment exclusions to sub-deployments.  This is the initial commit.

### CODE REVIEW

**Comment by bgaisford** [comment]
I'm still working on test cases.

**Comment by bstansberry** [comment]
Brandon, here are some details on the "Core - Full Integration Build 1170 outcome was FAILURE" reported above:

1) org.jboss.as.test.clustering.cluster.sso.ClusteredSingleSignOnTestCase(SYNC-tcp).testNoAuthSingleSignOn

Ignore this; it's a test that intermittently fails; no evidence it's related to your patch.

2) org.jboss.as.test.integration.messaging.mgmt.ConnectionFactoryManagementTestCase.testWriteDiscoveryGroupAttributeWhenConnectorIsAlreadyDefined

Not sure about this one, but it failed recently in another test run unrelated to this PR, so for now let's assume it's unrelated.

3) org.jboss.as.test.integration.deployment.classloading.ear.EarJbossStructureAdditionalModuleTestCase

Related. Here's the most informative part of the server.log output when the test was running:

```
19:45:58,611 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-4) WFLYSRV0027: Starting deployment of "e9c563ab-5c6f-45b1-8c7a-314ce3f40ab3.ear" (runtime-name: "e9c563ab-5c6f-45b1-8c7a-314ce3f40ab3.ear")
&amp#27;[0m&amp#27;[31m19:45:58,620 ERROR [org.jboss.msc.service.fail] (MSC service thread 1-4) MSC000001: Failed to start service jboss.deployment.unit."e9c563ab-5c6f-45b1-8c7a-314ce3f40ab3.ear".STRUCTURE: org.jboss.msc.service.StartException in service jboss.deployment.unit."e9c563ab-5c6f-45b1-8c7a-314ce3f40ab3.ear".STRUCTURE: WFLYSRV0153: Failed to process phase STRUCTURE of deployment "e9c563ab-5c6f-45b1-8c7a-314ce3f40ab3.ear"
    at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:163)
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.startService(ServiceControllerImpl.java:1948)
    at org.jboss.msc.service.ServiceControllerImpl$StartTask.run(ServiceControllerImpl.java:1881)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
    at org.jboss.as.server.deployment.module.descriptor.JBossDeploymentStructureParser13.parseModuleStructureSpec(JBossDeploymentStructureParser13.java:337)
    at org.jboss.as.server.deployment.module.descriptor.JBossDeploymentStructureParser13.parseSubDeployment(JBossDeploymentStructureParser13.java:298)
    at org.jboss.as.server.deployment.module.descriptor.JBossDeploymentStructureParser13.readElement(JBossDeploymentStructureParser13.java:252)
    at org.jboss.as.server.deployment.module.descriptor.JBossDeploymentStructureParser13.readElement(JBossDeploymentStructureParser13.java:68)
    at org.jboss.staxmapper.XMLMapperImpl.processNested(XMLMapperImpl.java:110)
    at org.jboss.staxmapper.XMLMapperImpl.parseDocument(XMLMapperImpl.java:69)
    at org.jboss.as.server.deployment.module.descriptor.DeploymentStructureDescriptorParser.parse(DeploymentStructureDescriptorParser.java:329)
    at org.jboss.as.server.deployment.module.descriptor.DeploymentStructureDescriptorParser.parse(DeploymentStructureDescriptorParser.java:307)
    at org.jboss.as.server.deployment.module.descriptor.DeploymentStructureDescriptorParser.deploy(DeploymentStructureDescriptorParser.java:160)
    at org.jboss.as.server.deployment.DeploymentUnitPhaseService.start(DeploymentUnitPhaseService.java:156)
    ... 5 more
```

**Comment by bgaisford** [comment]
I don't intend to do any further development on this PR within core.  As this PR contains v1.2 -> v1.3 copied+updated files to support the jboss-deployment-structure.xml schema update, it would be great to get these changes in master as soon as possible in order to avoid future merge conflicts that will arise if developers work in parallel on the old v1.2 files.  I intend to write integration tests for this feature within full wildfly.

File: server/src/main/resources/schema/jboss-deployment-structure-1_3.xsd [hunk 1]
```diff
@@ -0,0 +1,778 @@
...
+                        This is
+                        as per spec*
+                    </documentation>
+                </annotation>
+            </xsd:element>
+            <xsd:element name="ear-exclusions-cascaded-to-subdeployments" type="xsd:boolean" minOccurs="0" maxOccurs="1">
+                <annotation xmlns="http://www.w3.org/2001/XMLSchema">
+                    <documentation>
+                        Flag indicating whether .ear exclusions defined in the root deployment should be cascaded to
+                        sub-deployments.
```
Comment (bstansberry) [inline server/src/main/resources/schema/jboss-deployment-structure-1_3.xsd]: This only applies to the "exclusions" element in deploymentType and not to the "exclude-subsystems". Perhaps say "Flag indicating whether .ear module exclusions..."?

If someone reads the exclude-subsystems doc, its clear that this wouldn't apply, but it took me a while to figure that out.

**Comment by bstansberry** [comment]
If you get a chance to tweak the doc comment by tomorrow, great; if not I'll get this merged anyway. It's no big deal and Stuart and I have both looked at this and it looks good. Thanks!

**Comment by bgaisford** [comment]
@bstansberry  Thanks for the feedback, I updated the comments as per your request.  Hopefully that's more clear.  Please let me know if you think any further changes are needed.

**Comment by bstansberry** [comment]
Thanks!

**Pull request merged**